### ISSUE TITLE
Kubernetes Priority Annotation Doesn't Work

### ISSUE THREAD

**Issue by rawkode** [containous/traefik#2225]
### Do you want to request a *feature* or report a *bug*?
Bug

### What did you do?

Configure ingress as such:

```yml
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: api-service-ingress-api-gateway
  namespace: myapp
  annotations:
    kubernetes.io/ingress.class: traefik
    traefik.frontend.priority: 2
spec:
  tls:
  - hosts:
    - api.myapp.com
    secretName: api-tls
  rules:
  - host: api.myapp.com
    http:
      paths:
        - backend:
            serviceName: api-service
            servicePort: 10000
```
### What did you expect to see?

Priority applied to rule

### What did you see instead?

Priority 1 and no anotation listed under `kubectl describe`

### Output of `traefik version`: (_What version of Traefik are you using?_)

`1.4`

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Kubernetes / Helm Chart

**Comment by timoreimann** [comment]
This should work to the best of my knowledge and from taking a quick look at the code.

I'll run some tests locally to verify.

**Comment by timoreimann** [comment]
Wild guess: try specifying the priority as `"2"` instead of `2`?

**Comment by rawkode** [comment]
@timoreimann Let me run it again with the quotes and I'll get back to you

**Comment by rawkode** [comment]
@timoreimann It made no difference I'm afraid

**Comment by timoreimann** [comment]
@rawkode thanks for checking. I'll dig into the matter either tonight or some time tomorrow.

**Comment by timoreimann** [comment]
In the meantime, if you have some interesting DEBUG level logs to share, I'd be curious to see.

**Comment by timoreimann** [comment]
The feature is working as intended for me: I took the examples from [our guide](https://docs.traefik.io/user-guide/kubernetes/) and added the priority annotation to the Traefik UI Ingress object. The priority value was then reflected in the `/api` endpoint.

I did notice though that the priority value **must be quoted**: Missing that, it will be interpreted as a number and ignored by the API server since annotations must always be strings. Trying to `kubectl apply` such an ill-formated will simply drop the illegal annotation; with `kubectl edit` (or `kubectl patch`), you'd actually see this error message:

```
error: ingresses "traefik-web-ui" could not be patched: cannot convert int64 to string
```

Could you double-check again that you specified the annotation correctly? If you are really certain, then I'd need a reproducible set of manifests (Deployment/Pod, Ingress, Service) for further analysis.

**Comment by rawkode** [comment]
Thanks for investigating, @timoreimann. I'll delete the resources and force a refresh and see if this helps.

Thanks again

**Comment by rawkode** [comment]
Still can't get this to work. `kubectl` shows the annotations, but Traefik doesn't do anything with them.

![screenshot from 2017-10-09 18-07-57](https://user-images.githubusercontent.com/145816/31349903-eedeab88-ad1c-11e7-920b-079e79fef12e.png)

![screenshot from 2017-10-09 18-08-21](https://user-images.githubusercontent.com/145816/31349909-f26c16e6-ad1c-11e7-849c-0135aab268bb.png)


Traefik is installed via the Helm chart, version `1.3.8`

```
# Ingress
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: app-user-service-ingress
  namespace: app
  annotations:
    kubernetes.io/ingress.class: traefik
    traefik.frontend.priority: "20"
```

```
# Ingress
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: app-api-ingress
  namespace: app
  annotations:
    kubernetes.io/ingress.class: traefik
    traefik.frontend.priority: "40"
```

**Comment by rawkode** [comment]
@timoreimann If it helps, I'll explain what I'm trying to-do:

I have 3 services that each response to different paths on `api.app.com`. One of the services is the "catch-all".

So I'm trying to increase the priority of the two that have specific paths, such as `/login` and `/register` and then finally match on the catch-all.

**Comment by timoreimann** [comment]
Priority and weight are two different things. Could you please check the output of the `/api` endpoint on Traefik's web page port?

**Comment by rawkode** [comment]
@timoreimann It seems to have assigned some routes a priority of its own

```json
{
  "kubernetes": {
    "backends": {
      "api.app.com": {
        "servers": {
          "api-service-deployment-385312239-nqh58": {
            "url": "http://10.52.3.120:10000",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      },
      "api.app.com/login": {
        "servers": {
          "user-service-deployment-599394805-wblnc": {
            "url": "http://10.52.2.68:4000",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      },
      "api.app.com/register": {
        "servers": {
          "user-service-deployment-599394805-wblnc": {
            "url": "http://10.52.2.68:4000",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      },
      "api.app.com/verify-email": {
        "servers": {
          "user-service-deployment-599394805-wblnc": {
            "url": "http://10.52.2.68:4000",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      },
      "app.app.com": {
        "servers": {
          "web-deployment-2081565715-6rgwl": {
            "url": "http://10.52.1.125:80",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      },
      "traefik.app.com": {
        "servers": {
          "traefik-traefik-2185114437-fltjl": {
            "url": "http://10.52.2.85:8080",
            "weight": 1
          }
        },
        "loadBalancer": {
          "method": "wrr"
        }
      }
    },
    "frontends": {
      "api.app.com": {
        "entryPoints": ["http", "https"],
        "backend": "api.app.com",
        "routes": {
          "api.app.com": {
            "rule": "Host:api.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 0,
        "basicAuth": []
      },
      "api.app.com/login": {
        "entryPoints": ["http", "https"],
        "backend": "api.app.com/login",
        "routes": {
          "/login": {
            "rule": "PathPrefix:/login"
          },
          "api.app.com": {
            "rule": "Host:api.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 6,
        "basicAuth": []
      },
      "api.app.com/register": {
        "entryPoints": ["http", "https"],
        "backend": "api.app.com/register",
        "routes": {
          "/register": {
            "rule": "PathPrefix:/register"
          },
          "api.app.com": {
            "rule": "Host:api.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 9,
        "basicAuth": []
      },
      "api.app.com/verify-email": {
        "entryPoints": ["http", "https"],
        "backend": "api.app.com/verify-email",
        "routes": {
          "/verify-email": {
            "rule": "PathPrefix:/verify-email"
          },
          "api.app.com": {
            "rule": "Host:api.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 13,
        "basicAuth": []
      },
      "app.app.com": {
        "entryPoints": ["http", "https"],
        "backend": "app.app.com",
        "routes": {
          "app.app.com": {
            "rule": "Host:app.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 0,
        "basicAuth": []
      },
      "traefik.app.com": {
        "entryPoints": ["http", "https"],
        "backend": "traefik.app.com",
        "routes": {
          "traefik.app.com": {
            "rule": "Host:traefik.app.com"
          }
        },
        "passHostHeader": true,
        "priority": 0,
        "basicAuth": []
      }
    }
  }
}
```

**Comment by timoreimann** [comment]
By default Traefik uses the path length for frontends coming with a path matcher. That's why you're seeing non-zero priorities on Ingresses that specify paths.

Let me dig a bit deeper on my end now that I know your concrete use cases. Thanks!

**Comment by rawkode** [comment]
Thank you. I appreciate you looking in to this!

It's been driving me crazy

**Comment by timoreimann** [comment]
@rawkode I'm having a hunch here... are you possibly still running Traefik 1.3 even though you mentioned 1.4 in your bug report?

I can totally not reproduce the problem with the latest version, but the picture matches perfectly with 1.3, and my local tests confirm that.

**Comment by rawkode** [comment]
@timoreimann I started on 1.4, then switched to 1.3.8, as I thought 1.4 was the problem :joy:

**Comment by rawkode** [comment]
Thanks for your help, upgrading to 1.4 again does indeed fix the priorities.

Notes for anyone else with similar problems:

I switched to 1.3.8 because the weight didn't change and I wasn't aware of the difference and because `kubectl describ` doesn't list annotations (Thought they are available in `kubectl edit`)

**Comment by timoreimann** [comment]
@rawkode glad we figured it out! 1.3 did not support specifying priorities through annotations yet, that feature was only shipped along 1.4.

The output of `kubectl describe` on annotations does look like it's missing something: I see an `Annotations:` field but no content even though my Ingress contains an annotation. Could be a bug.
